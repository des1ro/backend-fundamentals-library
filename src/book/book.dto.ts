export class Book {
  constructor(
    readonly title: string,
    readonly author: string,
    readonly year: number
  ) {}
}
