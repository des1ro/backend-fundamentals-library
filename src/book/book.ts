export class Book {
  constructor(
    readonly title: string,
    readonly author: string,
    readonly year: number
  ) {}
}
/**
 * book ->
 *  book.service.ts
 *  book.dto.ts
 *
 *
 * book ->
 *    services ->
 *      book.service.ts
 *      newspaper.service.ts
 *    dto ->
 *      book.dto.ts
 *       .....
 */
