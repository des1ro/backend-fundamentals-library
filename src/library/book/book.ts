export class Book {
  constructor(
    private readonly title: string,
    private readonly author: string,
    private readonly year: number,
    private readonly id: number
  ) {}
  getId(): number {
    return this.id;
  }
  getTitle(): string {
    return this.title;
  }
  getAuthor(): string {
    return this.author;
  }
  getYear(): number {
    return this.year;
  }
}
