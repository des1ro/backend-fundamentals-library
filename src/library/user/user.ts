import { randomUUID } from "crypto";

export class User {
  private readonly uuid: string = randomUUID();
  constructor(
    private readonly name: string,
    private penaltyPoints: number = 0,
    private dueDate = new Date()
  ) {}
  getName(): string {
    return this.name;
  }
  getUuid(): string {
    return this.uuid;
  }
  getPenaltyPoints() {
    return this.penaltyPoints;
  }
  addPenaltyPoints(penaltyPoints: number): void {
    this.penaltyPoints += penaltyPoints;
  }
  getDueDate() {
    return this.dueDate;
  }
  setDueDate(date: Date) {
    this.dueDate = date;
  }
  resetPenaltyPoints() {
    this.penaltyPoints = 0;
  }
}
