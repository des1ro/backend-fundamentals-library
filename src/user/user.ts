import { randomUUID } from "crypto";
import { isAfter } from "date-fns";
export class User {
  readonly uuid: string = randomUUID();
  private dueDate?: number;
  private penaltyPoints = 0;
  constructor(readonly name: string) {}
  getPenaltyPoints() {
    return this.penaltyPoints;
  }
  addPenaltyPoints(penaltyPoints: number): void {
    this.penaltyPoints += penaltyPoints;
  }
  subtractPenaltyPoints(penaltyPoints: number): void {
    this.penaltyPoints -= penaltyPoints;
  }
  getDueDate() {
    return this.dueDate || 0;
  }
  setDueDate(date: number): void {
    this.dueDate = date;
  }
  isBanned(): boolean {
    return isAfter(this.getDueDate(), new Date());
  }
}
