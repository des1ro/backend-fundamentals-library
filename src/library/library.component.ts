import { Book } from "../book/book.dto";
import { LibraryError } from "../exceptions/library.exceptions";
import { User } from "../user/user.dto";
import { Booking } from "../booking/booking.service";
import { UsersUpdate } from "../user/service/usersUpdate/usersUpdate.service";
import * as cron from "node-cron";

export class Library {
  private readonly usersUpdate: UsersUpdate;
  constructor(
    private readonly libraryBooks = new Map<Book, number>(),
    private readonly bookings = new Map<User, Booking>()
  ) {
    this.usersUpdate = new UsersUpdate();
  }
  constantlyUpdateUsers(): void {
    cron.schedule("* 30 5 * * * ", () => {
      console.log("Update users every day at 5:30");
      this.usersUpdate.update(this.bookings);
      console.log("Library users updated");
    });
  }
  addBook(book: Book, amount = 1): void {
    const value = this.libraryBooks.get(book) || 0;
    this.libraryBooks.set(book, amount + value);
  }
  deleteBook(book: Book, amount: number = 1): void {
    const currentAmount = this.libraryBooks.get(book);
    if (!currentAmount) {
      throw new LibraryError({
        name: "INVALID_BOOK",
        message: "The book is not available in this library",
      });
    }
    const newAmount = currentAmount - amount;
    if (newAmount > 0) {
      this.libraryBooks.set(book, newAmount);
      return;
    }
    this.libraryBooks.delete(book);
  }
  createBooking(user: User) {
    if (this.bookings.has(user)) {
      throw new LibraryError({
        name: "BOOKING_ERROR",
        message: "Booking already exist",
      });
    }
    this.bookings.set(user, new Booking(user));
  }
  borrowBook(user: User, book: Book) {
    const booking = this.bookings.get(user);
    const currentAmount = this.libraryBooks.get(book) || 0;
    if (currentAmount === 0) {
      throw new LibraryError({
        name: "INVALID_BOOK",
        message: "Book isn't in library",
      });
    }
    if (booking) {
      this.libraryBooks.set(book, currentAmount - 1);
      booking.addBookToUser(book);
      return;
    }
    throw new LibraryError({
      name: "BOOKING_ERROR",
      message: "Booking doen't exist",
    });
  }
  returnBook(user: User, book: Book) {
    const booking = this.bookings.get(user);
    if (!booking) {
      throw new LibraryError({
        name: "INVALID_BOOK",
        message: "Book not found",
      });
    }
    const currentAmount = this.libraryBooks.get(book);
    if (currentAmount !== undefined) {
      booking.returnBook(book);
      this.libraryBooks.set(book, currentAmount + 1);
      return;
    }
    throw new LibraryError({
      name: "INVALID_BOOK",
      message: "Book not found in library",
    });
  }
}
