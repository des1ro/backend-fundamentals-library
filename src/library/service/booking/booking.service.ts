import { isAfter, addMonths } from "date-fns";
import { Book } from "../../../book/book.dto";
import { LibraryError } from "../../../exceptions/library.exceptions";
import { User } from "../../../user/user.dto";

export class Booking {
  private readonly penaltyPointsPerBook: number;
  constructor(
    private readonly user: User,
    private readonly booksBorrowDate = new Map<Book, Date>()
  ) {
    this.penaltyPointsPerBook = 10;
  }
  addBookToUser(book: Book): void {
    const isBanned = isAfter(this.user.getDueDate(), new Date());
    if (isBanned) {
      throw new LibraryError({
        name: "PERMISION_DENIED",
        message: "User is banned!",
      });
    }
    this.booksBorrowDate.set(book, new Date());
  }
  returnBook(returnedBook: Book): void {
    const date = this.booksBorrowDate.get(returnedBook);
    if (!date) {
      throw new LibraryError({
        name: "INVALID_BOOK",
        message: "The book is not in the user's list",
      });
    }
    const bookBorrowDateLimit = addMonths(date, 1);
    const bookIsAfterDueDate = isAfter(new Date(), bookBorrowDateLimit);
    this.booksBorrowDate.delete(returnedBook);
    if (bookIsAfterDueDate) {
      const penaltyPoints =
        this.user.getPenaltyPoints() + this.penaltyPointsPerBook;
      this.user.setPenaltyPoints(penaltyPoints);
    }
  }
  getBooksDatesArray(): Date[] {
    const datesArray = Array.from(this.booksBorrowDate.values());
    return datesArray;
  }
}
