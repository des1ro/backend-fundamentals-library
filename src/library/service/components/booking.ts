import { add, addMonths, differenceInMonths, isAfter } from "date-fns";
import { Book } from "../../book/book";
import { User } from "../../user/user";
import { LibraryError } from "../../exceptions/library.exceptions";
// sprawdzać przy oddawaniu czy książka po terminie
// wszystko w bookingu
// user ma mieć
export class Booking {
  constructor(
    private readonly user: User,
    private readonly booksDueDate = new Map<Book, Date>()
  ) {}
  getUser(): User {
    return this.user;
  }
  isUserBanned(
    bookBorrowMonthsLimit: number,
    penaltyPointsLimit: number
  ): boolean {
    const isBanned = isAfter(this.user.getDueDate(), new Date());
    if (isBanned) {
      return true;
    }
    const userPenaltyPoints = this.checkUserPenaltyPoints(
      bookBorrowMonthsLimit,
      penaltyPointsLimit
    );
    if (userPenaltyPoints > 0) {
      const banMonths = userPenaltyPoints / 10;
      this.user.setDueDate(addMonths(new Date(), banMonths));
      return true;
    }
    return false;
  }
  addBookToUser(
    book: Book,
    bookBorrowMonthsLimit: number,
    penaltyPointsLimit: number
  ): void {
    if (!this.isUserBanned(bookBorrowMonthsLimit, penaltyPointsLimit)) {
      this.booksDueDate.set(book, new Date());
      return;
    }
    throw new LibraryError({
      name: "PERRMISION_DENIED",
      message: "User is banned!",
    });
  }
  checkUserPenaltyPoints(
    bookBorrowMonthsLimit: number,
    penaltyPointsPerBook: number
  ): number {
    let penaltyPoints = 0;
    this.booksDueDate.forEach((bookDate) => {
      const difference = differenceInMonths(new Date(), bookDate);
      if (difference >= bookBorrowMonthsLimit)
        penaltyPoints += penaltyPointsPerBook;
    });
    this.user.addPenaltyPoints(penaltyPoints);
    return this.user.getPenaltyPoints();
  }

  returnBook(
    returnedBook: Book,
    bookBorrowMonthsLimit: number,
    penaltyPointsPerBook: number
  ): void {
    const isUserBanned = this.isUserBanned(
      bookBorrowMonthsLimit,
      penaltyPointsPerBook
    );
    if (this.booksDueDate.has(returnedBook)) {
      const date = this.booksDueDate.get(returnedBook);
      if (date && isAfter(new Date(), addMonths(date, 1)) && isUserBanned) {
        this.user.addPenaltyPoints(-penaltyPointsPerBook);
      }
      this.booksDueDate.delete(returnedBook);
      return;
    }
    throw new LibraryError({
      name: "INVALID_BOOK",
      message: "The book is not in the user's list",
    });
  }
  getBooksDueDate(): Map<Book, Date> {
    return this.booksDueDate;
  }
  getUserDueDate() {
    return this.user.getDueDate();
  }
  getUserPenaltyPoints() {
    return this.user.getPenaltyPoints();
  }
}
