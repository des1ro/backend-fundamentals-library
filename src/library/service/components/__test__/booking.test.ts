import { addDays, addMonths } from "date-fns";
import { Book } from "../../../book/book";
import { User } from "../../../user/user";
import { Booking } from "../booking";
import { LibraryError } from "../../../exceptions/library.exceptions";
jest.mock("../../../book/book");
describe("Booking test suite", () => {
  let objectUnderTest: Booking;
  let mockedUser: User;
  let mockedBookOne: Book;
  let mockedBookTwo: Book;
  let testDate: Date;
  beforeAll(() => {
    mockedBookOne = new Book("Test title One", "Test author One", 1992, 1);
    mockedBookTwo = new Book("Test title Two", "Test author Two", 1995, 2);
  });
  beforeEach(() => {
    mockedUser = new User("Test user");
    jest.useFakeTimers();
  });
  afterEach(() => {
    jest.clearAllMocks();
    jest.useRealTimers();
  });
  it("Should get user properly", () => {
    //Given
    objectUnderTest = new Booking(mockedUser);
    const user = objectUnderTest.getUser();
    //Then
    expect(user).toBe(mockedUser);
  });
  it("Should get dueData properly", () => {
    //Given
    testDate = new Date();
    objectUnderTest = new Booking(mockedUser, testDate);
    const date = objectUnderTest.getDueDate();
    //Then
    expect(date).toBe(testDate);
  });
  it("Should check if book is in booking", () => {
    //Given
    objectUnderTest.borrowBook(mockedBookTwo);
    //When
    const resultOne = objectUnderTest.checkBook(mockedBookOne);
    const resultTwo = objectUnderTest.checkBook(mockedBookTwo);
    //Then
    expect(resultOne).toBeFalsy();
    expect(resultTwo).toBeTruthy();
  });
  describe("Get user credit test suite", () => {
    it("Should increase credit if books reach borrow limit", () => {
      //Given
      testDate = addDays(new Date(), 30);
      const borrowLimit = 25;
      objectUnderTest = new Booking(mockedUser);
      const spy = jest.spyOn(objectUnderTest, "borrowBook");
      //When
      objectUnderTest.borrowBook(mockedBookOne);
      jest.setSystemTime(testDate);
      const result = objectUnderTest.getUserCredit(borrowLimit);
      const expectedResult = 30 - 25;
      //Then
      expect(spy).toBeCalledTimes(1);
      expect(result).toBe(expectedResult);
    });
    it("Should get user credit if books don't reach borrow limit", () => {
      //Given
      const testBookBorrowLimit = 1;
      objectUnderTest = new Booking(mockedUser);
      //When
      objectUnderTest.borrowBook(mockedBookOne);
      objectUnderTest.borrowBook(mockedBookTwo);
      const result = objectUnderTest.getUserCredit(testBookBorrowLimit);
      const expectedResult = mockedUser.getCredit();
      //Then
      expect(result).toBe(expectedResult);
    });
    it("Should reset the user's credit if it is blocked", () => {
      //Given
      const testBookBorrowLimit = 20;
      testDate = addDays(new Date(), 30);
      mockedUser.setCredit(5);
      objectUnderTest = new Booking(mockedUser, testDate);
      const spy = jest.spyOn(objectUnderTest, "getUserCredit");
      //When
      const result = mockedUser.getCredit();
      const resultCredit = objectUnderTest.getUserCredit(testBookBorrowLimit);
      //Then
      expect(result).toBe(5);
      expect(resultCredit).toBe(0);
      expect(spy).toBeCalledTimes(1);
    });
  });
  it("Should block user for one month", () => {
    //Given
    objectUnderTest = new Booking(mockedUser);
    const spy = jest.spyOn(objectUnderTest, "getBlockUser");
    //When
    const expectedResult = addMonths(new Date(), 1);
    objectUnderTest = objectUnderTest.getBlockUser();
    //Then
    expect(spy).toBeCalledTimes(1);
    expect(objectUnderTest.getDueDate()).toStrictEqual(expectedResult);
  });
  it("Should return book properly", () => {
    //Given
    jest.setSystemTime(new Date());
    objectUnderTest = new Booking(mockedUser);
    const spy = jest.spyOn(objectUnderTest, "returnBook");
    //When
    objectUnderTest.borrowBook(mockedBookOne);
    const resultOne = objectUnderTest.checkBook(mockedBookOne);
    objectUnderTest.returnBook(mockedBookOne);
    const resultTwo = objectUnderTest.checkBook(mockedBookOne);
    //Then
    expect(resultOne).toBeTruthy();
    expect(resultTwo).toBeFalsy();
    expect(spy).toBeCalledWith(mockedBookOne);
  });
  it("Should throw an error on returing wrong book", () => {
    //Given
    objectUnderTest = new Booking(mockedUser);
    const spy = jest.spyOn(objectUnderTest, "checkBook");
    //When
    objectUnderTest.borrowBook(mockedBookOne);
    const resultOne = objectUnderTest.checkBook(mockedBookOne);
    const resultTwo = objectUnderTest.checkBook(mockedBookTwo);
    //Then
    expect(resultOne).toBeTruthy();
    expect(resultTwo).toBeFalsy();
    expect(spy).toBeCalledTimes(2);
    expect(() => objectUnderTest.returnBook(mockedBookTwo)).toThrow(
      LibraryError
    );
  });
});
