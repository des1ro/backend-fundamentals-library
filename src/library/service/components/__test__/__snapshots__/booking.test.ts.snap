// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Booking test suite Return book test suite Should return the book without any changes if it was returned on time 1`] = `Map {}`;

exports[`Booking test suite Return book test suite Should return the book without any changes if it was returned on time 2`] = `Map {}`;

exports[`Booking test suite Should add book to user properly 1`] = `
Map {
  Book {
    "author": "Test author One",
    "id": 1,
    "title": "Test title One",
    "year": 1992,
  } => 2023-07-20T10:30:00.000Z,
}
`;
